optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/Joiners.java

```diff
@@ -151,6 +152,23 @@
     // join(..., planningVariableContainsCached(Talk::getPeriod, (Period a, Period b) -> a.overlaps(b)))
     // get the period value range, does a cartesian product on it, so it maps every period to an overlapping periodList
     // then keep an index from every period to all talks in an overlapping period (possible the same period)
+    public static <A, B, Property_ extends Comparable<Property_>> BiJoiner<A, B> overlapsOrMeets(
+            Function<A, Property_> leftStartMapping,
+            Function<A, Property_> leftEndMapping,
+            Function<B, Property_> rightStartMapping,
+            Function<B, Property_> rightEndMapping) {
+        return AbstractBiJoiner.merge(lessThanOrEqual(leftStartMapping, rightEndMapping),
+                greaterThanOrEqual(leftEndMapping, rightStartMapping));
+    }
+
+    public static <A, B, Property_ extends Comparable<Property_>> BiJoiner<A, B> overlaps(
+            Function<A, Property_> leftStartMapping,
+            Function<A, Property_> leftEndMapping,
+            Function<B, Property_> rightStartMapping,
+            Function<B, Property_> rightEndMapping) {
```

Comment:
Did you consider any other way of specifying the intervals?
(I'm not saying this one is bad. Just making sure we went over all the options before we picked a public API to be with us forever.)

Reply:
An other way would be `startA, startB, endA, endB`
When generalizing this to higher cardinality, it looks like this:
```
overlaps((a,b) -> a.start, c -> c.start, (a,b) -> a.end, c -> c.end)
```

Meanwhile, with the current definition, it be generalized like this:
```
overlaps((a,b) -> a.start, (a,b) -> a.end, c -> c.start, c -> c.end)
```

I see these as the two main contenders if we use a Function -> Comparable API for the following reasons:
1. The other joiners have the "main" stream as the first argument (this eliminates (start2,  ...) (end2, ...))
2. The main convention I see in other code is either (start1,end1,start2,end2)  (first interval followed by second interval) or  (start1,start2,end1,end2) (start points followed by end points)

The alternative way of defining it is to have a "Duration" (ex: 2 hour shift that start at 3am), but that requires an additional method to "add" a point and the point must still be comparable, so I don't see any advantage to that approach.

Reply:
The idea of `(startA, durationA, startB, durationB)` is intriguing. Not as the only option, but as an alternative API alongside the `(start1,end1,start2,end2)`. The former can just delegate to the latter. Why?

In order to support a wider range of data models:
- People who have a data model with start date and job length, they use the former, without having to have complex lambdas like `a -> // add duration to the start date`.
- People who have a data model with a start/end dates, they use the former.

A win-win, I'd say.

Reply:
That requires the duration to be of type `Duration`. Some people might use int, doubles etc. for durations. So we need an 'add' method for this to be generic, or we need to create a Temporal, Duration special case.

Reply:
I think it's a safe bet that, if they have `startDate`  as `OffsetDateTime`, they'd have a job length as `Duration`. Otherwise, if it's just an `int` for example, they're making life harder for themselves - and we'll support that anyway. So yes, I do think we should have a specialization for `Duration`. (I think we already do that in `ConstraintCollectors`.)